### PR TITLE
Update factors to return residual

### PR DESCRIPTION
**Pull request by Affie**

### CODE REVIEW

File: src/factors/Bearing2D.jl [hunk 1]
```diff
@@ -40,12 +37,11 @@ function (cfo::CalcFactor{<:Pose2Point2Bearing})( res::AbstractVector{<:Real},
   reuse.resid .-= reuse.predvec
 
   # must return result of length zDim==1 in this case
-  res[1] = sum(abs.(reuse.resid))
+  return sum(abs.(reuse.resid))
```
Comment (dehann) [inline src/factors/Bearing2D.jl:40]: we will probably drop the reuse.residual container here at some point, right?

File: src/factors/BearingRange2D.jl [hunk 2]
```diff
@@ -51,57 +44,27 @@ function (cfo::CalcFactor{<:Pose2Point2BearingRange})(res::AbstractVector{<:Real
...
-
-  eθ = atan((my + xi[2]), (mx + xi[1])) - atan(lm[2], lm[1])
-
-  res = [eθ, er]
-
-  return res' * iΣ * res
-
-end
-
-
```
Comment (dehann) [inline src/factors/BearingRange2D.jl:104]: this is great thanks!  consolidated.

File: src/factors/DynPoint2D.jl [hunk 3]
```diff
@@ -24,21 +24,16 @@ mutable struct DynPoint2DynPoint2{T <: SamplableBelief} <: AbstractRelativeRoots
...
+
+function (cfo::CalcFactor{<:DynPoint2DynPoint2})(z, xi, xj)
   #
   dt = Dates.value(cfo.metadata.fullvariables[2].nstime - cfo.metadata.fullvariables[1].nstime)*1e-9   # roughly the intended use of userdata
-  res[1:2] = z[1:2] - (xj[1:2] - (xi[1:2]+dt*xi[3:4]))
-  res[3:4] = z[3:4] - (xj[3:4] - xi[3:4])
-  nothing
+  res12 = z[1:2] - (xj[1:2] - (xi[1:2]+dt*xi[3:4]))
+  res34 = z[3:4] - (xj[3:4] - xi[3:4])
+  return [res12; res34]
```
Comment (dehann) [inline src/factors/DynPoint2D.jl:33]: we'll see how the allocations go.  if bad then we can turn to using the cfo.residual container -- likely best done after consolidating CPT?

File: src/factors/Range2D.jl [hunk 4]
```diff
@@ -13,17 +13,13 @@ Point2Point2Range(d::D) where {D <: IIF.SamplableBelief} = Point2Point2Range{D}(
...
-            theta,
-            xi,
-            lm )
+function (cfo::CalcFactor{<:Point2Point2Range})(rho, theta, xi, lm)
   #
   XX = lm[1] - (rho[1]*cos(theta[1]) + xi[1])
   YY = lm[2] - (rho[1]*sin(theta[1]) + xi[2])
-  res[1] = XX^2 + YY^2
-  return nothing
+  #TODO JT - Should this have a sqrt for parametric?
```
Comment (dehann) [inline src/factors/Range2D.jl:20]: my feeling is yes, else the Malahanobis distance will be wrong.  else factors will be mixing ^2 and ^4 costs.  sort is a little excess but safest path for now i think

File: src/factors/VelPose2D.jl [hunk 5]
```diff
@@ -16,27 +16,50 @@ VelPose2VelPose2(z1::T1, z2::T2) where {T1 <: IIF.SamplableBelief, T2 <: IIF.Sam
...
+
+  return meas, iΣ
+end
+
+function (cf::CalcFactor{<:VelPose2VelPose2})(meas,
                                               Xi,
                                               Xj  )
   #
+  #FIXME JT - Createing new res for simplicity, it may not hold up well though
+  res = Vector{eltype(Xi)}(undef, 5)
```
Comment (dehann) [inline src/factors/VelPose2D.jl:39]: ah cool okay, let's try a few ways and that will hopefully help show how to do it natively in cfo.residual for all factors, available if users want to use it to reduce allocations.  Think it's okay to experiment with this for a bit

File: test/testBearingRange2D.jl [hunk 6]
```diff
@@ -122,23 +122,23 @@ res = testFactorResidualBinary( p2br,
...
-@test isapprox(f([0,0,pi/2], [-10.,0]), 0, atol = 1e-9)
-
-f = Pose2Point2BearingRange(Normal(pi,1), Normal(10.0,1))
-@test isapprox(f([0.,0,0], [-10.,0]), 0, atol = 1e-9)
-@test isapprox(f([0,0,pi/2], [0.,-10]), 0, atol = 1e-9)
-
-f = Pose2Point2BearingRange(Normal(-pi/2,1), Normal(10.0,1))
-@test isapprox(f([0.,0,0], [0.,-10]), 0, atol = 1e-9)
-@test isapprox(f([0,0,pi/2], [10.,0]), 0, atol = 1e-9)
+# #TODO Update to new CalcFactor
```
Comment (dehann) [inline test/testBearingRange2D.jl:125]: yeah, i've been trying to find a nice API in IIf for these tests.  I basically start with IIF.testResidualBinary (or near about)

File: test/testParametric.jl [hunk 7]
```diff
@@ -160,4 +162,31 @@ vardict, result, varIds, Σ = IIF.solveFactorGraphParametric(fg, useCalcFactor=t
...
+
+addVariable!(fg, :x1, DynPose2;  nanosecondtime=1000_000_000)
+
+# conditional likelihood between Dynamic Point2
+dp2dp2 = VelPose2VelPose2(MvNormal([10.0;0;0], [0.01;0.01;0.001]), MvNormal([0.0;0], [0.1; 0.1]))
+addFactor!(fg, [:x0;:x1], dp2dp2)
+
+ensureAllInitialized!(fg)
+
+vardict, result, varIds, Σ = IIF.solveFactorGraphParametric!(fg)
```
Comment (dehann) [inline test/testParametric.jl:187]: very cool!

**Review by dehann — APPROVED** [review]
all good thanks!  Let's tag so long but then experiment with different ways to get the right type on cfo.residual?  The user can then find a memory allocation problem in factor function and hook into the (latest possible "internally" pre-allocated) cfo.residual container?

File: src/factors/Bearing2D.jl (hunk 1 above)
Comment (Affie) [reply src/factors/Bearing2D.jl:40]: Yes, there are a couple of different ways in use currently. I agree we can experiment and then standardise.

File: src/factors/DynPoint2D.jl (hunk 3 above)
Comment (Affie) [reply src/factors/DynPoint2D.jl:33]: The function will probably get inlined, so it shouldn't make much of a difference. NLSolve reuses the residual in any case.

File: src/factors/Range2D.jl (hunk 4 above)
Comment (Affie) [reply src/factors/Range2D.jl:20]: See this comparison from slack: "Actually the getSample theta part (2*pi*rand(N)) won't work in any case"
https://github.com/JuliaRobotics/RoME.jl/compare/feat/21Q1/factor_residuals...twig/21Q1/parametricPoint2Point2Range?expand=1

File: test/testParametric.jl (hunk 7 above)
Comment (Affie) [reply test/testParametric.jl:187]: It just takes some time to update and test, but most factors should work with AD without much effort.

**Pull request merged**